File: q1_two_sum/cpp/solution1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, vector<int> > numsidx;
        vector<int> delidx;
        vector<int> rtnidx;
        int count = nums.size();
        for (int i = 0; i < count; i++) {
            int num = nums[i];
            numsidx[num].push_back(i);
            if (numsidx[num].size() == 2) {
                if (2 * num != target) {
                    delidx.push_back(num);
                } else {
                    rtnidx = numsidx[num];
                    return rtnidx;
                }
            }
        }

        for (auto &delnum : delidx) {
            numsidx.erase(delnum);
        }

        for (auto &num : numsidx) {
            int delta = target - num.first;
            if (numsidx.find(delta) != numsidx.end()) {
                if (num.second[0] != numsidx[delta][0]) {
                    rtnidx = {num.second[0], numsidx[delta][0]};
                    return rtnidx;
                }
            }
        }
    }
};

File: q1_two_sum/py/solution1.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        dict = {}
        delidx = []
        for i in range(len(nums)):
            num = nums[i]
            try:
                dict[num].append(i)
                if 2 * num != target:
                    delidx.append(num)
                else:
                    return dict[num]
            except KeyError:
                dict[num] = [i]

        for num in set(delidx):
            del dict[num]

        for key in dict.keys():
            delta = target - key
            try:
                if dict[key][0] != dict[delta][0]:
                    return [dict[key][0], dict[delta][0]]
            except KeyError:
                pass
